Disable IPv6 in the default installation

tgraf opened the pull request:
This standardizes the enablement option to `--enable-ipv4` and `--enable-ipv6` while preserving the old `--disable-ipv4` for backward compatibility.

It changes both the Docker and CNI plugin to support with IPv6 mode disabled. The plugins will refuse to provide service if neither address is provided.

It makes all IPv6 code in the datapath optional so it can be disabled.

Finally, the PR disabled IPv6 for all new deployments while preserving the existing behavior for existing deployments. This reduces the memory footprint in default installations significantly.

Depends on: #6080
Fixes: #1537 

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/6082)
<!-- Reviewable:end -->

joestringer commented on Documentation/cmdref/cilium-agent.md:
```diff
@@ -37,9 +37,10 @@ cilium-agent
-      --disable-ipv4                                Disable IPv4 mode
       --disable-k8s-services                        Disable east-west K8s load balancing by cilium
   -e, --docker string                               Path to docker runtime socket (DEPRECATED: use container-runtime-endpoint instead) (default "unix:///var/run/docker.sock")
+      --enable-ipv4                                 Enable IPv4 support (default true)
```
(commit title: `Univeral` -> `Universal`)

joestringer commented on bpf/bpf_lxc.c:
```diff
@@ -707,15 +712,19 @@ int handle_ingress(struct __sk_buff *skb)
 		ep_tail_call(skb, CILIUM_CALL_IPV4_FROM_LXC);
 		ret = DROP_MISSED_TAIL_CALL;
 		break;
+#endif
```
Consider extending the `#endif` to include ARP, given that ARP is used for IPv4 neighbour resolution?

joestringer commented on daemon/daemon.go:
```diff
@@ -591,15 +597,17 @@ func (d *Daemon) init() error {
+				}
+				if err := lbmap.RRSeq6Map.DeleteAll(); err != nil {
+					return err
+				}
```
Does it make sense for the cleanup to be specific to whether IPv4/IPv6 is enabled? This cleanup functionality is in relation to a previous Cilium run, and we don't know whether the previous run had IPv4/IPv6 enabled, so if someone deploys Cilium with standard DS then disables eg IPv6 then the maps will never be cleared?

joestringer replied:
Looks like the final commit takes care of this concern.

joestringer commented on daemon/state.go:
```diff
@@ -325,22 +325,24 @@ func (d *Daemon) regenerateRestoredEndpoints(state *endpointRestoreState) {
-	}(ep)
+
+		defer func(ep *endpoint.Endpoint) {
+			if err != nil {
```
For this defer function to work as intended, the same `err` variable needs to be shared across the entire function.

aanm replied:
@tgraf @joestringer that's not enough, the returned error needs to have the variable `err` assigned.

```
func (d *Daemon) allocateIPsLocked(ep *endpoint.Endpoint) (err error) {
....
```

tgraf replied:
@aanm Why? I don't think that is true at all.

aanm replied:
@tgraf https://play.golang.org/p/IjBet0lGfY9

tgraf replied:
I see what you mean but we only need err to be non-nil which will be the case. The defer to clean up is called when necessary.

joestringer commented on pkg/ipcache/ipcache.go:
```diff
@@ -288,7 +288,7 @@ func (ipc *IPCache) Upsert(ip string, hostIP net.IP, newIdentity Identity) bool
 		}
 	} else {
-		scopedLog.Error("Attempt to upsert invalid IP into ipcache layer")
+		scopedLog.Panic("Attempt to upsert invalid IP into ipcache layer")
```
This change in log level is surprising given that it's causing a panic in a commit that doesn't explicitly call this out. Is this intentional?

tgraf replied:
Good catch, this was a leftover from debugging.

joestringer reviewed (dismissed):
Looks good overall.

A couple of notes:
* The deferred release of allocated IPv6 address below doesn't look like it works correctly in error conditions
* It'd be nice to teach `cilium-health` that the IPv4 and/or IPv6 addresses are not expected to work, so that it doesn't always complain about lack of connectivity. (Could be a follow-up)

aanm commented on daemon/state.go:
```diff
@@ -325,26 +325,27 @@ func (d *Daemon) regenerateRestoredEndpoints(state *endpointRestoreState) {
+	var err error
 
-	defer func(ep *endpoint.Endpoint) {
+	if option.Config.EnableIPv6 && ep.IPv6 != nil {
```
If the endpoint didn't contain an IPv6 shouldn't we try to assign a new one now?

tgraf replied:
That's potentially quite dangerous as the application inside the container is already running. Assigning an IPv6 address in the middle may break things. Just like we keep using addresses when disabling an address family, we should probably not add a new address to already existing containers.

raybejjani reviewed (dismissed):
Looks good except for the `allocateIPsLocked` err handling.

tgraf committed:
cni: Do not require IPv6 addressing information to be present

Fail if neither IPv4 and IPv6 addressing is available

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
agent: Remove BPF maps of disabled address families

This ensures that memory resources are freed up when disabling an address
family.

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
agent: Universal --enable-ipv4 and --enable-ipv6 option

The old --disable-ipv4 is marked as hidden but kept for backwards
compatibility. Map the DISABLE_IPV4 environment variable to --enable-ipv4 so
existing ConfigMap remain compatible.

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
agent: Fail if both IPv4 and IPv6 are disabled

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
kubernetes: Disable IPv6 by default for new deployments

The agent continues to enable IPv6 in the default setting but the default
ConfigMap will disable IPv6 for all new deployments. The motivation is
primarily a smaller memory footprint in default deployments.

This reduces the memory footprint in the default installation by about 200M.

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
agent: Make IPv6 optional

Datapath:
Use a consistent ENABLE_IPV4 and ENABLE_IPV6 define to steer enabling IPv4 and
IPv6 code paths. Make all IPv6 datapath code dependent on ENABLE_IPV6.

Agent:
Make installation of routes, creation of maps, insertion into ipcache, etc.
dependent on option.Config.EnableIPv6.

Also fixes a problem in bpf/Makefile to properly test a combination of defines
when performing test compilation.

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
docker: Remove the unneeded EndpointGet check

It only delays the setup process and adds no value. Docker is the source of
truth for docker endpoint IDs.

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
docker: Only install routes if address family is enabled

This allows to disable IPv4 and IPv6 for docker networks. Only one of the two
address families have to be enabled.

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
ipam: Allow disabling IPv6 in node addressing API

A flag already existed to indicate enablement but IPv6 was hardwired to
enabled.  Use the --enable-ipv6 value to control the node addressing API.

Signed-off-by: Thomas Graf <thomas@cilium.io>

aanm reviewed (approved):
I'm glad the default IPv6 is still `true`
